message/validation/common_checks.go

```diff
@@ -0,0 +1,118 @@
+		e := ErrLateSlotMessage
+		e.got = fmt.Sprintf("late by %v", lateness)
+		return e
+	}
+
+	return nil
+}
+
+// messageEarliness returns how early message is or 0 if it's not
+func (mv *messageValidator) messageEarliness(slot phase0.Slot, receivedAt time.Time) time.Duration {
+	return mv.netCfg.Beacon.GetSlotStartTime(slot).Sub(receivedAt)
+}
+
+// messageLateness returns how late message is or 0 if it's not
+func (mv *messageValidator) messageLateness(slot phase0.Slot, role spectypes.RunnerRole, receivedAt time.Time) time.Duration {
+	var ttl phase0.Slot
+	switch role {
+	case spectypes.RoleProposer, spectypes.RoleSyncCommitteeContribution:
+		ttl = 1 + lateSlotAllowance
+	case spectypes.RoleCommittee, spectypes.RoleAggregator:
+		ttl = phase0.Slot(mv.netCfg.Beacon.SlotsPerEpoch()) + lateSlotAllowance
+	case spectypes.RoleValidatorRegistration, spectypes.RoleVoluntaryExit:
+		return 0
+	}
+
+	deadline := mv.netCfg.Beacon.GetSlotStartTime(slot + ttl).
+		Add(lateMessageMargin)
+
+	return receivedAt.Sub(deadline)
+}
+
+func (mv *messageValidator) validateDutyCount(
+	msgID spectypes.MessageID,
+	msgSlot phase0.Slot,
+	validatorIndexCount int,
+	signerStateBySlot *OperatorState,
+) error {
+	dutyCount := signerStateBySlot.DutyCount(mv.netCfg.Beacon.EstimatedEpochAtSlot(msgSlot))
+
+	dutyLimit, exists := mv.dutyLimit(msgID, msgSlot, validatorIndexCount)
+	if !exists {
+		return nil
+	}
+
+	if dutyCount >= dutyLimit {
+		err := ErrTooManyDutiesPerEpoch
+		err.got = fmt.Sprintf("%v (role %v)", dutyCount, msgID.GetRoleType())
+		err.want = fmt.Sprintf("less than %v", dutyLimit)
+		return err
+	}
+
+	return nil
+}
+
+func (mv *messageValidator) dutyLimit(msgID spectypes.MessageID, slot phase0.Slot, validatorIndexCount int) (int, bool) {
+	switch msgID.GetRoleType() {
+	case spectypes.RoleVoluntaryExit:
+		pk := phase0.BLSPubKey{}
+		copy(pk[:], msgID.GetDutyExecutorID())
+
+		return mv.dutyStore.VoluntaryExit.GetDutyCount(slot, pk), true
+
+	case spectypes.RoleAggregator, spectypes.RoleValidatorRegistration:
+		return 2, true
+
+	case spectypes.RoleCommittee:
+		return 2 * validatorIndexCount, true
+
+	default:
+		return 0, false
+	}
+}
+
+func (mv *messageValidator) validateBeaconDuty(
+	role spectypes.RunnerRole,
+	slot phase0.Slot,
+	indices []phase0.ValidatorIndex,
+) error {
+	// Rule: For a proposal duty message, we check if the validator is assigned to it
+	if role == spectypes.RoleProposer {
+		epoch := mv.netCfg.Beacon.EstimatedEpochAtSlot(slot)
+		if mv.dutyStore.Proposer.ValidatorDuty(epoch, slot, indices[0]) == nil {
+			return ErrNoDuty
+		}
+	}
+
+	// Rule: For a sync committee aggregation duty message, we check if the validator is assigned to it
+	if role == spectypes.RoleSyncCommitteeContribution {
+		period := mv.netCfg.Beacon.EstimatedSyncCommitteePeriodAtEpoch(mv.netCfg.Beacon.EstimatedEpochAtSlot(slot))
+		if mv.dutyStore.SyncCommittee.Duty(period, indices[0]) == nil {
+			return ErrNoDuty
+		}
+	}
```

Comment:
@MatheusFranco99 @nkryuchkov 

Why we use `indices[0]` here? how do we know this is the correct validator index from the whole committee validator list?

Reply:
this should probably be checked with either all the indices or the right index that really has duty.

Reply:
In this function, I think it's assumed that, if the role is either Proposer or Sync Committee Contribution, the indices list will only have one element. Right @nkryuchkov ?

Reply:
It's checked somewhere else but we can add here an extra check to see that the length of the list is 1, for independence, if you like

Reply:
https://github.com/ssvlabs/ssv/pull/1496

Reply:
its not what I meant, I didn't see the part where we are passing the 1 len list. I was now able to track it down to the code at `getCommitteeAndValidatorIndices` where we populate the list with only one validator when the message belongs to a propose duty.